custom/icds/view_utils.py

```diff
@@ -23,3 +26,18 @@ def _inner(request, domain, *args, **kwargs):
 
 def is_icds_cas_project(domain):
     return IS_ICDS_ENVIRONMENT and domain == ICDS_DOMAIN
+
+
+def check_app_access(domain, user, app):
+    if (
+        settings.SERVER_ENVIRONMENT == 'icds-staging'
+        and toggles.ROLE_WEBAPPS_PERMISSIONS.enabled(domain)
+    ):
+        try:
+            role = user.get_role(domain)
+        except DomainMembershipError:
+            return 'User is not a member of this project', 404
+        else:
+            if not (role and role.permissions.view_web_app(app)):
```

Comment:
I would vote to change this api to pass in the app ID instead of the entire app. There are only a few places this is called and if you have the app you can call `view_web_app(app.master_id)`.

Doing that would prevent having to load the app at all.

Reply:
Hey @snopoke 
Looking at the API here 
https://github.com/dimagi/commcare-hq/blob/ap/restrict-unauthorized-users/corehq/apps/users/models.py#L172

I am not sure of it, do you think it we can drop `app['copy_of']`?
 
If yes, does the following snippet looks good?
```python
def view_web_app(self, app_id):
    if self.view_web_apps:
        return True
    return app_id in self.view_web_apps_list
```

Reply:
yes, that's correct. `self.view_web_apps_list` should only contain IDs of 'master' apps and not builds. By checking `app['_id']` and `app['copy_of']` the method could be called with either an app build or a master app. With your changes it can only be called with a 'master' app so callers must use `app.master_id` if they have the app or else ensure that the ID being passed in is the ID of a master app and not a build.

In the case of heartbeat we have both master and build IDs so we much just make sure we pass in the correct one (the one from the request GET params.

Reply:
What do you think would be a good way to go about it?
Create a new function that takes in app_id or update the existing function and all the places where it is called?
It is being used in three other places on HQ. 

Should I create a different PR if I am updating the existing function?

Reply:
Hey @AmitPhulera 
Its only called at ~3 places, so should be okay to edit this method in this PR itself. It's anyway needed to roll out this change, to avoid load on redis

Reply:
Okay. Cool. 
I will update it. Thanks 😄